src/Stackage/Package/Git/Types.hs

```diff
@@ -0,0 +1,211 @@
+    -- ^ Directory name with a trailing slash, without it trees wouldn't get
+    -- sorted properly: @("0.1" > "0.1.1") /= ("0.1/" > "0.1.1/")@
+  deriving (Show)
+
+
+type TreePath = [FileName]
+
+
+-- | Convert file path to a tree path.
+toTreePath :: FilePath -> TreePath
+toTreePath path = foldr toFileName [] splitPath'
+  where
+    splitPath' = S8.split '/' $ S8.pack path
+    toFileName fName []
+      | null fName =
+        error $ "Tree path should end with a file name, not a directory: " ++ path
+      | otherwise = [FileName $ BS.toShort fName]
+    toFileName fDir tp = DirectoryName (BS.toShort $ S8.snoc fDir '/') : tp
+
+
+instance Eq FileName where
+  (==) (FileName f1) (FileName f2) = f1 == f2
+  (==) (FileName f) (DirectoryName d) = BS.fromShort f == S8.init (BS.fromShort d)
+  (==) (DirectoryName d) (FileName f) = S8.init (BS.fromShort d) == BS.fromShort f
+  (==) (DirectoryName d1) (DirectoryName d2) = d1 == d2
+
+
+instance Ord FileName where
+  compare (FileName f1) (FileName f2) = compare f1 f2
+  compare f@(FileName fn) d@(DirectoryName dn)
+    | f == d = EQ
+    | otherwise = compare fn dn
+  compare d@(DirectoryName dn) f@(FileName fn)
+    | f == d = EQ
+    | otherwise = compare dn fn
+  compare (DirectoryName d1) (DirectoryName d2) = compare d1 d2
+
+-- | A very tight representation of SHA1 hash.
+data ShortRef =
+  ShortRef !Word64
+           !Word64
+           !Word32
+  deriving (Eq, Ord)
+
+
+-- | Reduce memory usage by squeezing `Ref` into a much more memory efficient
+-- representation `ShortRef`.
+toShortRef :: MonadIO m => Ref -> m ShortRef
+toShortRef !(toBinary -> bs)
+    | length bs /= 20 = error "Stackage.Package.Git.Types.toShortRef: length is not 20"
+    | otherwise = liftIO $ BU.unsafeUseAsCString bs $ \ptr -> ShortRef
+        <$> peekByteOff ptr 0
+        <*> peekByteOff ptr 8
+        <*> peekByteOff ptr 16
+
+-- | Convert `ShortRef` back into a `Ref`, so it can be used with Hit package.
+fromShortRef :: ShortRef -> Ref
+fromShortRef !(ShortRef w64 w64' w32) =
+  fromBinary $
+  L.toStrict $
+  B.toLazyByteString
+    (BE.word64Host w64 <> BE.word64Host w64' <> BE.word32Host w32)
+
+
+-- | All possible file types expected by git.
+data FileType
+  = ExecFile
+    -- ^ Executable - @0o100755@
+  | NonExecFile
+    -- ^ Normal file - @0o100644@
+  | NonExecGroupFile
+    -- ^ Grandfathered group writeable mode - @0o100664@ - @git fsck@ allows
+    -- this permissions:
+    -- <https://github.com/git/git/blob/8354fa3d4ca50850760ceee9054e3e7a799a4d62/fsck.c#L583>
+  | SymLink
+    -- ^ Symbolic link - @0o120000@
+  | GitLink
+    -- ^ Git link, used for modules - @0o160000@
+  deriving (Eq, Show)
+
+
+-- | A recursive git tree that can either be used to keep current representation
+-- of tree or simulate an in-memory work tree.
+data WorkTree d f
+  = File !f !FileType
+  | Directory !d !(Map.Map FileName (WorkTree d f))
+  deriving (Show)
+
+
+instance Eq (WorkTree ShortRef ShortRef) where
```

Comment:
There is no need to descend into all subtrees and check equality on those, which would be the case in derived implementation. Here on directories it only checks equality of SHAs, which means all subtrees are also equal: `(Directory ref1 _) == (Directory ref2 _) = ref1 == ref2`